[Merged by Bors] - Add `AttesterCache` for attestation production

paulhauner opened the pull request:
## Issue Addressed

- Resolves #2169

## Proposed Changes

Adds the `AttesterCache` to allow validators to produce attestations for older slots. Presently, some arbitrary restrictions can force validators to receive an error when attesting to a slot earlier than the present one. This can cause attestation misses when there is excessive load on the validator client or time sync issues between the VC and BN.

## Additional Info

NA

paulhauner committed:
Add first compiling draft

paulhauner committed:
Refactor to remove enum

paulhauner committed:
Fix lint, ensure cache is primed

paulhauner committed:
Add max len to cache

paulhauner committed:
Add some initial comments

paulhauner committed:
Change cache key to decision root

paulhauner committed:
Add comments

paulhauner committed:
Remove Vec from cache

paulhauner committed:
Fix bug with cache key, add comments

paulhauner committed:
Handle uninitialized cache, add comments

paulhauner committed:
Add comments

paulhauner committed:
Tidy beacon chain

paulhauner committed:
Apply restriction to HTTP api

paulhauner committed:
Tidy use of Option, add comments

paulhauner committed:
Tidy comments

paulhauner committed:
Fix HTTP error

paulhauner committed:
Add explicit error for ancient slot

paulhauner committed:
Remove restriction from API

paulhauner committed:
Remove unused variable

paulhauner committed:
Tidy, add comments

paulhauner committed:
Move state reads into AttesterCache

paulhauner committed:
Add comments for cache size

paulhauner committed:
Add warn log for cache miss

paulhauner committed:
Tidy comments

michaelsproul commented on beacon_node/beacon_chain/src/attester_cache.rs:
```diff
@@ -0,0 +1,380 @@
+//!
+//! This cache is required *as well as* the `ShufflingCache` since the `ShufflingCache` does not
+//! provide any information about the `state.current_justified_checkpoint`. It is not trivial to add
+//! the justified checkpoint to the `ShufflingCache` since that cache keyed by shuffling decision
```
```suggestion
//! the justified checkpoint to the `ShufflingCache` since that cache is keyed by shuffling decision
```

michaelsproul commented on beacon_node/beacon_chain/src/attester_cache.rs:
```diff
@@ -0,0 +1,380 @@
+    }
+}
+
+/// Stores the minimal amount of data required compute the committee length for any committee at any
```
```suggestion
/// Stores the minimal amount of data required to compute the committee length for any committee at any
```

michaelsproul commented on beacon_node/beacon_chain/src/attester_cache.rs:
```diff
@@ -0,0 +1,380 @@
+    ///
+    /// The `latest_block_root` should be the latest block that has been applied to `state`. This
+    /// parameter is required since the state does not store the block root for any block with the
+    /// same slot as `slot.slot()`.
```
```suggestion
    /// same slot as `state.slot()`.
```

michaelsproul commented on beacon_node/beacon_chain/src/attester_cache.rs:
```diff
@@ -0,0 +1,380 @@
+    }
+}
+
+impl Eq for AttesterCacheKey {}
```
Deriving should work, no?

michaelsproul commented on beacon_node/beacon_chain/src/attester_cache.rs:
```diff
@@ -0,0 +1,380 @@
+    ///
+    /// This function takes a write-lock on the internal cache. Prefer attempting a `Self::get` call
+    /// before running this function as `Self::get` only takes a read-lock and is therefore less
+    /// likely to create head contention.
```
```suggestion
    /// likely to create contention.
```

michaelsproul commented on beacon_node/beacon_chain/src/attester_cache.rs:
```diff
@@ -0,0 +1,380 @@
+                .map(|(key, _)| key)
+                .min_by_key(|key| key.epoch)
+                // Only return values whilst the cache is full.
+                .filter(|_| cache.len() >= MAX_CACHE_LEN)
```
I doubt we'll hit this much in practice (due to the cache being oversized), but seeing as this is `O(n)` I think we may as well move the `filter` before the `min_by_key` and do just 1 iteration instead of 2.

paulhauner replied:
I refactored this a bit more to ensure it only does the minimum iterations. Thanks for this, I hadn't considered the extra iter.

michaelsproul commented on beacon_node/beacon_chain/src/beacon_chain.rs:
```diff
@@ -1217,44 +1220,168 @@ impl<T: BeaconChainTypes> BeaconChain<T> {
+                warn!(
+                    self.log,
+                    "Attester cache miss";
+                    "beacon_block_root" => %beacon_block_root,
```
Thoughts on showing the full block root to save archaeological work? 

```suggestion
                    "beacon_block_root" => ?beacon_block_root,
```

michaelsproul commented on beacon_node/beacon_chain/src/beacon_chain.rs:
```diff
@@ -1217,44 +1220,168 @@ impl<T: BeaconChainTypes> BeaconChain<T> {
+                // Permit attesting to slots *prior* to the current head. This is desirable when
+                // the VC and BN are out-of-sync due to time issues or overloading.
+                beacon_block_root = *head_state.get_block_root(request_slot)?;
+                beacon_state_root = *head_state.get_state_root(request_slot)?;
```
Technically we could speed this up by fetching the state at the previous epoch boundary, although I'm wary of over-complicating it for marginal benefit. Maybe just something to keep in mind if this cases comes up a lot in practice (it shouldn't!)

paulhauner replied:
I agree, I had considered this but avoided it to save complexity. It'll be an easy fix if we decide we need it later.

paulhauner replied:
I ended up implementing this because I was kinda forced to deal with the complexity to deal with cloning the head state :slightly_smiling_face:

michaelsproul commented on beacon_node/beacon_chain/src/beacon_chain.rs:
```diff
@@ -2086,6 +2214,17 @@ impl<T: BeaconChainTypes> BeaconChain<T> {
             }
         }
 
+        // Apply the state to the attester cache, only if it is from the previous epoch or earlier.
```
This seems mismatched with the condition, which adds states with `epochs >= current_epoch - 2`, i.e. newer than the previous previous epoch.

michaelsproul commented on beacon_node/beacon_chain/src/attester_cache.rs:
```diff
@@ -0,0 +1,380 @@
+        }
+
+        let mut state: BeaconState<T::EthSpec> = chain
+            .get_state(&state_root, None)?
```
I think we should (continue to) use `chain.state_at_slot` here, to avoid re-reading the state from the DB in the case where the state advance timer fails to prime the cache. One case where this will happen is if there are two consecutive skipped slots either side of the epoch boundary (slot 31 and slot 0 mod 32).

paulhauner replied:
`BeaconChain::state_at_slot` is racey since it takes another read-lock on the head without checking to see if the head is still relevant to the desired `state_root`.

I'd avoided cloning the head for this reason, but I just thought of a fairly tidy way to do it in 8cb488efe5cd9fdd628bc6b7edb0cffd2ed980b7. Let me know what you think. 

> (continue to)

I'm not sure what you mean by this. With the present code we were never using `state_at_slot`.

paulhauner replied:
Actually, I need to this about this a bit more. I'll ping you when I'm happy with it.

paulhauner replied:
Alright, I spent some more time with this and I ended up and in a similar position as before. I'll do my best to explain:

There's a property I want to maintain with this cache, let's call it "no double loads". Lets say a BN has enough validators to span all committee indices and they all hit `validator/attestation_data` at exactly the same time. If there is a cache miss, I want to avoid the scenario where we load/clone a `BeaconState` for each of those committees (i.e., 64x). I want the only *first* request to trigger a state load and to block all the other requests on that. You probably saw comments about this in `AttesterCache::load_and_cache_state`. 

Another property I want to maintain is "no lock interleaving". I.e., don't hold the `chain.canonical_head` and `chain.attester_cache.cache` locks at the same time.

Unfortunately, I can't see a simple and efficient solution where we maintain these properties *and* avoid the DB read by cloning the head.

I think we could get away with interleaving with the present code-base, but I'm concerned about future regressions. I'm interested in your thoughts.

> One case where this will happen is if there are two consecutive skipped slots either side of the epoch boundary (slot 31 and slot 0 mod 32).

My current thinking is that if we start to get burned by this then we should make the state advance timer keep advancing the state for each skipped slot (rather than just advancing once). I can't recall why we don't already do this, but it seems like a good idea to me :thinking:

paulhauner replied:
> My current thinking is that if we start to get burned by this then we should make the state advance timer keep advancing the state for each skipped slot (rather than just advancing once). I can't recall why we don't already do this, but it seems like a good idea to me thinking

I remembered. It's because we need to store intermediate states if we use that state to import a block.

michaelsproul commented on beacon_node/beacon_chain/src/beacon_chain.rs:
```diff
@@ -1217,44 +1220,168 @@ impl<T: BeaconChainTypes> BeaconChain<T> {
+                // This scenario is likely to happen occasionally. It is reasonable to drop this to
+                // a `debug!` if it turns out to be frequent and unavoidable. At least whilst the
+                // feature is new it is nice to see if the cache is getting a lot of misses.
+                warn!(
```
I think this should be `debug!` due to the skipped slots case

michaelsproul reviewed (changes requested):
Good stuff!

Just a few minor changes requested

paulhauner committed:
Add metrics

paulhauner committed:
Apply suggestions from code review

Co-authored-by: Michael Sproul <micsproul@gmail.com>

paulhauner committed:
Partially address review comments

paulhauner committed:
Clone the head state

paulhauner committed:
Fix state root, load target

paulhauner commented:
I've addressed your comments, but one of the changes turned out to be a little complex. I'm going to go do something else and then come back and do another self-review with a fresh mind.

paulhauner committed:
Revert "Fix state root, load target"

This reverts commit cefab045e0618d46d0bc8a2319ad80ace49b8d56.

paulhauner committed:
Revert "Clone the head state"

This reverts commit 8cb488efe5cd9fdd628bc6b7edb0cffd2ed980b7.

paulhauner commented:
I've done some more thinking and left some more comments here: https://github.com/sigp/lighthouse/pull/2478#discussion_r678126154

I don't see a perfect solution to avoid the DB read at this point. I'll sit with it overnight and see how I feel about it tomorrow morning.

paulhauner commented:
Just to be explicit, this PR represents the following change in the scenario where both epoch-boundary-adjacent slots are skipped:

### Present Behavior

For each `validator/attestation_data` request (`O(n)`, where `n` is usually between 0-64):
 - Clone the head state
 - Perform per-slot/per-epoch processing
 
 ### New Behavior
 
 Assuming no reorgs, only once (`O(1)`):
 
 - Read the DB from disk
 - Perform per-slot/per-epoch processing

michaelsproul reviewed (approved):
Ok, yeah. I agree with your assessment of the situation. I think the code as it stands now (e7c3008124db18f28a03f0f9a981b258173cd7a9) strikes a good compromise between avoiding clones _and_ DB reads, while also avoiding data races. Although it would be nice to avoid the DB read when we could theoretically clone from the head, the complications involved in doing that appear not to be worth it, and the solution proposed in 8cb488efe5cd9fdd628bc6b7edb0cffd2ed980b7 is too costly as it potentially clones the head for every request (30ms each or more).

I'd be happy to merge this as-is, pending more positive data from Prater

paulhauner commented:
> I'd be happy to merge this as-is, pending more positive data from Prater

This has been running without issue on Prater for a few days. It's been running in the current form for ~24hrs without issue as well.

I'll merge it in now, we'll end up doing more testing with v1.5.0 release testing too. Thanks for the review!

bors r+
